Support independent includes with conditional, optional, and remote entries

tldahlgren opened the pull request:
Closes #40018.
Closes #31026.
Closes #2700.

There were a number of feature requests for os-specific config. This enables os-specific config without adding a lot of special sub-scopes.

Supersedes #46792. (Technically it starts with the same set of commits as that PR.)

Support `include:` as an independent configuration schema, allowing users to include
configuration scopes from files or directories. Includes can be:
* conditional (similar to definitions in environments), and/or
* optional (i.e., the include will be skipped if it does not exist).

Includes can be paths or URLs (`ftp`, `https`, `http` or `file`). Paths can be absolute or
relative . Environments can include configuration files using the same schema. Remote includes 
must be checked by `sha256`.

Includes can also be recursive, and this modifies the config system accordingly so that
we push included configuration scopes on the stack *before* their including scopes, and
we remove configuration scopes from the stack when their including scopes are removed.

For example, you could have an `include.yaml` file (e.g., under `$HOME/.spack`) to specify
global includes:

```
include:
- ./enable_debug.yaml
- path: https://github.com/spack/spack-configs/blob/main/NREL/configs/mac/config.yaml
  sha256: 37f982915b03de18cc4e722c42c5267bf04e46b6a6d6e0ef3a67871fcb1d258b
```

Or an environment `spack.yaml`:

```
spack:
  include:
  - path: "/path/to/a/config-dir-or-file"
    when: os == "ventura"
  - ./path/relative/to/containing/file/that/is/required
  - path: "/path/with/spack/variables/$os/$target"
    optional: true
  - path: https://raw.githubusercontent.com/spack/spack-configs/refs/heads/main/path/to/required/raw/config.yaml
    sha256: 26e871804a92cd07bb3d611b31b4156ae93d35b6a6d6e0ef3a67871fcb1d258b
```

Updated TODO:
- [x] Get existing unit tests to pass with Todd's changes
- [x] Resolve new (or old) circular imports
- [x] Ensure remote includes (global) work
- [x] Ensure remote includes for environments work (note: caches remote
      files under user cache root)
- [x] add sha256 field to include paths, validate, and require for remote includes
- [x] add sha256 remote file unit tests
- [x] revisit how diamond includes should work
- [x] support recursive includes
- [x] add recursive include unit tests
- [x] update docs and unit test to indicate ordering of recursive includes with
      conflicting options is deferred to follow-on work

scheibelp committed:
Initial pass to support config:include:
 - partial include logic as subsection of config: (scheibelp 01f69fb17e60da973c3d9402817b554123ddf891)
 - ref error (scheibelp 161097ab3198049f56312de8fbdcae8e461b01ca)
 - refactor function used only for env encludes to support importing other lists of paths as config scopes (scheibelp e3b0da247fd48989b584bcd9bf36f063ea14f6eb)
 - initial test for included configs (scheibelp ae1aaa9065ff61979edf3d04cc468967c1b8512c)
 - style fix (scheibelp a54e5651339d3d4f5ae39933dc079a8a5d7957ba)
 - expanded test with additional entries (scheibelp a6795af237492feb609d50fcc2962f3f428494ed)
 - add docstring (scheibelp a18952e2b19e48dbe6da0895bc8ca037ad60f968)
 - all commands now make use of config:includes (scheibelp bd666d82345fbd5ecfc6d19fd3938a1104eeb77f)
 - includes -> include; file -> path; also indicate in schema that additional properties are not allowed (scheibelp b9af793acf09de3bd0cb9c5b44940dfa4ef0f417)
 - allow include: entries to be strings if they don't have an associated 'when' (scheibelp 5899d453dd5a23fd8d0b2e5e5e7e629317d7cd83)
 - auto style fix (scheibelp c9bb265c61e6a195e4cb5f6856a64958da4e96e6)
 - avoid repetition in test logic (scheibelp fe27fb3665fcd5789987b24a9bc4dea153d24e30)
 - now that include: has a path-only option, 'when' is always required when specifying 'path' (scheibelp f635b8e906623fee56bc345082e83876132337f7)
 - enforce requiring 'path' and 'when' at the schema level (scheibelp 5e07209819f0e7b36cef4d2fd57e20ca9560bb2c)
 - now that all used properties must be mentioned, fix an unrelated test (scheibelp 47684509b38d67439138522d8d1a84afb06172e5)
 - rewrite test in terms of a config section that actually exists (scheibelp 3fae4d422df34d8cd112e61aa5fe931cc354c8be)
 - other test depending on (scheibelp made-up) config:info refactored (da281022f9f8764fc24821104271a0e1b5409c3d)
 - test was somehow passing before based on now-deprecated reference to install_tree (previously install_root) - failing now that I disallow made-up properties in config (scheibelp 8a04e0eaa0164ea5949c2dfd72cda21d7c32f591)
 - partial testing of canonicalized include paths and optional includes (scheibelp e376e9191bd6cb7f6c6f5437a58483963bb04340)
 - better testing of prior change; added explanatory comments to test (scheibelp X60671a2d8e1a18c90f91a511235bb9a08496059c)
 - auto style fix (scheibelp 6787336b68dfdaff3e33f6182d9d847c90bed651)
 - update_config_with_includes: fix section name to match change (tldahlgren 444819ba9c48c0836c1d2b48c41c177f8cd67668)
 - scopes_from_paths: added docstring and type hints; converted to using f-strings (tldahlgren ff5fe7b96b9a34f78751b656af127066981f8548)

tldahlgren committed:
Schema: separate 'include' into own section
- initial separation (tldahlgren 34d96c69b3dbdf192db621d3bf8018d973248b36)
- include schema: fix starting line in file's docstring (tldahlgren 83346673d6d079d3ab68ef030f3648c16c1123f5)
- Environment schema: (re)use new include schema (tldahlgren a303d5e735ce592c248bf6e8b97bcfd45e6bd9b4)

tldahlgren committed:
remove support for config:include
- initial removal (tldahlgren 296971c041a46223fba7ae7f29bbcf1ba57a1f62)
- Restore unrelated test bug fixed in ccb69cc (tldahlgren 887ec635cde2198292bf94548018dca99a695b11)

tldahlgren committed:
Environments: support the new include schema

tldahlgren committed:
Post-rebase cleanup

tldahlgren committed:
Docs/environment includes: update to include new optional and conditional examples

tldahlgren committed:
Fix style issue with env schema

tldahlgren committed:
Remove setting config's additionalProperties to False (per Todd)

tldahlgren committed:
Restore config::update_config_with_includes; move included_config_scopes, scopes_from_paths to config

tldahlgren committed:
A pass over circular dependencies

tldahlgren committed:
test_env_include_configs: tweak spack.yaml string

tldahlgren committed:
includes: update docs

tldahlgren committed:
Restore separate env and include processing to fix env tests

tldahlgren committed:
Rebase cleanup: same diffs as before the rebase

tldahlgren committed:
Post-rebase style updates

tldahlgren committed:
Move offending imports into update_config_with_includes (but still circular)

tldahlgren committed:
Fix style issues

tldahlgren committed:
Fix pylint issues with unspecified encoding

tldahlgren committed:
Moved scopes_from_paths from environment.py to config.py

tldahlgren committed:
Move update_config_with_includes from config to main

tldahlgren committed:
environment: remove TODO comment from included_config_scopes

tldahlgren committed:
scopes_from_paths: replace resolve_relative (callable) with relative_root

tldahlgren committed:
eval_conditional: move from environment to spec; make public function

tldahlgren committed:
Refactor to create config.included_path and add OptionalPath

tldahlgren committed:
Fix style issues

tldahlgren committed:
Update fish completion

tldahlgren committed:
config.scopes_from_paths: add docstring spacing for readthedocs

tldahlgren committed:
Update the copyright lines for the new files

tldahlgren committed:
schemas: restore literal lines since new copyright is one-liner

tldahlgren committed:
_concretize_together_where_possible: Remove commented out import

tldahlgren committed:
Replace scopes_from_paths with include_path_scope

tldahlgren committed:
main: remove unused import

tldahlgren committed:
Moved remote config retrieval out of config.py

tldahlgren committed:
Resolve style issues plus add include_path_scope return to docstring

tldahlgren committed:
test_config_change_existing: do NOT use normal misc cache path

tldahlgren committed:
Snapshot: added test_include_duplicate_source; some config cleanup

tldahlgren committed:
config: remove unused remote_path

tldahlgren committed:
test_env_with_included_config_precedence: fix docstring typo

tldahlgren committed:
Refactor retention and handling of sources associated with included paths

tldahlgren committed:
Cache web-specific config settings in util.web_config; move RemoteFileError

tldahlgren committed:
Cleanup prior; add remote_file_cache.py

tldahlgren committed:
Ignore type hint for passing config.CONFIG into web_config.update()

tldahlgren committed:
Move path processing from config to remote file utility

tldahlgren committed:
Fix readthedocs FetchError complaint

tldahlgren committed:
Restore environment include ordering lost in refactor

tldahlgren committed:
Add information on path options for global includes

tldahlgren committed:
test_updated_completion_scripts: change message to exclude global include scopes

tldahlgren committed:
Add missing colon to exception in docstring

tldahlgren committed:
Reverse global includes for precedence

tgamblin commented on lib/spack/docs/environments.rst:
```diff
@@ -657,24 +657,34 @@ This configuration sets the default compiler for all packages to
+
+Files are listed using paths to individual files or directories containing them.
+Path entries may be absolute or relative to the environment or specified as URLs.
+URLs to individual files need link to the *raw* form of the file's contents.
```
I'd just remove this line -- it seems obvious that the thing at the URL should be the file itself, and "the raw form" is a github/gitlab-specific thing.  If you want to keep it, I would put it as a note to users of GitHub and GitLab and link to [the docs](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-and-understanding-files#viewing-or-copying-the-raw-file-content) so people understand what it means.

tldahlgren replied:
It is currently important that the URL be to the raw form of the file so IMO it should be kept.

tgamblin replied:
like I said it should be a note -- or "raw" should be defined.  I don't think most users know what that means.

tgamblin commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -0,0 +1,45 @@
+
+Spack allows you to include configuration files through ``include.yaml``.
+Using the ``include:`` heading results in pulling in external configuration
+information to be used by any Spack command. The syntax is:
```
```suggestion
information to be used by any Spack command.
```

tgamblin commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -0,0 +1,45 @@
+shows all three. The first entry, ``/path/to/a/required/config.yaml``,
+indicates that included ``config.yaml`` file is required (so must exist).
+Use of ``optional: true`` for ``/path/to/$os/$target/config`` means
+the path is only applied if it exists. The condition ``os == "ventura"``
```
```suggestion
the path is only included if it exists. The condition ``os == "ventura"``
```

tgamblin commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -0,0 +1,45 @@
+Use of ``optional: true`` for ``/path/to/$os/$target/config`` means
+the path is only applied if it exists. The condition ``os == "ventura"``
+in the ``when`` clause for ``/path/to/os-specific/config-dir`` means the
+path is only applied when the operating system (``os``) is ``ventura``.
```
```suggestion
path is only included when the operating system (``os``) is ``ventura``.
```

tgamblin commented on lib/spack/docs/include_yaml.rst:
```diff
@@ -0,0 +1,45 @@
+URLs. URLs to individual files need to link to the *raw* form of the file's
+contents. Only the ``file``, ``ftp``, ``http`` and ``https`` protocols (or
+schemes) are supported. Spack-specific, environment and user path variables
+can be used. (See :ref:`config-file-variables` for more information.)
```
See note above about raw links.

```suggestion
Included files can be specified by path or by their parent directory.
Paths may be absolute, relative (to the configuration file including the path), 
or specified as URLs. Only the ``file``, ``ftp``, ``http`` and ``https`` protocols (or
schemes) are supported. Spack-specific, environment and user path variables
can be used. (See :ref:`config-file-variables` for more information.)
```

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -167,6 +172,12 @@ def get_section(self, section: str) -> Optional[YamlConfigDict]:
+            if section == "include" and data is not None:
+                # Save the sources of included configuration files to support
+                # relative paths.
+                CONFIG.add_include_source(path, data["include"], self.name)
```
This shouldn't be necessary. You can use `spack.util.path.canonicalize_path(path)` for any `syaml_str` that comes out of config -- e.g. `data["include"][0]`.  Whoever gets the path should do it like we do for other paths in config -- config system doesn't need to know.

Example:
```console
> cat ~/.spack/include.yaml
include:
  - ~/disable_verify.yaml
```

```console
> git diff
diff --git a/lib/spack/spack/config.py b/lib/spack/spack/config.py
index 83d4ec1d1d..e496d07423 100644
--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -177,6 +177,11 @@ def get_section(self, section: str) -> Optional[YamlConfigDict]:
                 # Save the sources of included configuration files to support
                 # relative paths.
                 CONFIG.add_include_source(path, data["include"], self.name)
+                print(path)
+                path_from_yaml = data["include"][0]
+                print("PATH:     ", path_from_yaml)
+                print("TYPE:     ", type(path_from_yaml))
+                print("CANONICAL:", spack.util.path.canonicalize_path(path_from_yaml))

         return self.sections[section]
```

```console
> spack config get repos
/Users/gamblin2/.spack/include.yaml
PATH:      ~/disable_verify.yaml
TYPE:      <class 'spack.util.spack_yaml.syaml_str'>
CANONICAL: /Users/gamblin2/disable_verify.yaml
repos:
- $spack/var/spack/repos/builtin
```

`canonicalize_path` should be using the `_start_mark` and `_end_mark` attrs on `syaml_str` (which are basically file/line data), but you don't have to care b/c it pulls them off the path string, gets the dirname, and joins.

tldahlgren replied:
I had forgotten to revisit this from the prior PR.

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -430,6 +447,10 @@ def __init__(self, *scopes: ConfigScope) -> None:
 
+        # Track configuration file source paths since not otherwise available
+        # once includes are processed.
+        self.includes_source: Dict[str, List[Tuple[str, str]]] = collections.defaultdict(list)
```
should be removed.

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -730,6 +763,53 @@ def print_section(self, section: str, blame: bool = False, *, scope=None) -> Non
+            if scope == "":
+                tty.warn(f"No scope name given for {source_path}, {include_data}")
+            entry = (scope or "", source_str)
+            self.includes_source[include_str].append(entry)
```
remove

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -730,6 +763,53 @@ def print_section(self, section: str, blame: bool = False, *, scope=None) -> Non
+            if scope is not None:
+                return os.path.dirname(source_path)
+
+        return None
```
remove -- this should be handled by `canonicalize_path`

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -774,13 +854,98 @@ def override(
 
+#: Data class for the relevance of an optional path conditioned on either
+#: a spec and or explicitly specified as optional.
+class OptionalPath(NamedTuple):
```
The name of this is confusing -- it sounds a lot like `Optional[Path]`, where `Optional` means "can be `None`, which isn't what you mean.  It's also only "optional" if `optional` is `True`.

I think this should just be called `IncludePath` since it represents a path in the `include:` section in config.

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -774,13 +854,98 @@ def override(
 
 
+#: Data class for the relevance of an optional path conditioned on either
+#: a spec and or explicitly specified as optional.
```
Is `when` a spec?  If this is supposed to be like [`when:` for definitions](https://spack.readthedocs.io/en/latest/environments.html#spec-list-references), it shouldn't be -- it's just a limited python expression.

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -279,7 +290,13 @@ def get_section(self, section: str) -> Optional[YamlConfigDict]:
+        if section == "include" and data is not None:
+            # Save the source of included configuration files to support
+            # relative paths.
+            CONFIG.add_include_source(self.path, data["include"], self.name)
```
shouldn't need this special case

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -774,13 +854,98 @@ def override(
+
+
+def included_path(entry: Union[str, dict]) -> OptionalPath:
+    """Convert the included path entry into a standard form.
```
I think "standard form" means an `OptionalPath` (or rather `IncludePath`, but I would try not to use words like "standard" b/c it's not clear exactly what it means (unless there's actually a standard somewhere, in which case I'd expect a link).

```suggestion
    """Convert the included path entry into an IncludePath.
```

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -891,6 +1056,11 @@ def add_from_file(filename: str, scope: Optional[str] = None) -> None:
+            if section == "include":
+                # Save the source of included configuration files to support
+                # relative paths.
+                CONFIG.add_include_source(filename, value, scope)
```
remove special case

tgamblin commented on lib/spack/spack/util/web_config.py:
```diff
@@ -0,0 +1,52 @@
+"""This module breaks the spack.config<->spack.util.web circular import.
+
+Usage: spack.main is expected to instantiate/update the WebConfig instance
+after configuration files are processed.
```
This seems like a very drastic change just to break a circular import, and now config logic is duplicated, and we have a large maintainability burden (keeping this thing updated).  I think this needs a re-think.  Config is a central piece of Spack and we shouldn't have two ways to do it.

tldahlgren replied:
I agree and figured this would be controversial. So I was expecting push-back.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -5199,6 +5199,13 @@ def get_host_environment() -> Dict[str, Any]:
     }
 
 
+def eval_conditional(string):
+    """Evaluate conditional definitions using restricted variable scope."""
+    valid_variables = get_host_environment()
+    valid_variables.update({"re": re, "env": os.environ})
+    return eval(string, valid_variables)
```
I'm not sure why this was moved here, as it has nothing to do with specs -- I think someplace in `spack.util` makes more sense.

tldahlgren replied:
I was removing circular imports when I moved it here. IIRC I tried a couple of alternatives but `get_host_environment()` was creating circular import challenges.

tgamblin commented on lib/spack/spack/test/cmd/commands.py:
```diff
@@ -260,14 +260,16 @@ def test_updated_completion_scripts(shell, tmpdir):
     msg = (
-        "It looks like Spack's command-line interface has been modified. "
-        "Please update Spack's shell tab completion scripts by running:\n\n"
+        "It looks like Spack's command-line interface has been\nmodified. "
+        "If differences are more than your global 'include:' scopes,\nplease "
+        "update Spack's shell tab completion scripts by running:\n\n"
```
I'm not sure what the extra `\n`'s are for here -- are they intended to make the text wrapping nicer for this message?  If so I would just run `msg` through `textwrap.wrap` instead of trying to insert `\n`'s manually.

tldahlgren replied:
Yes, they were to make it more readable on the traditional width window.

tgamblin replied:
Use `textwrap` to make this more clear.

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -43,10 +39,9 @@
 from spack import traverse
 from spack.installer import PackageInstaller
-from spack.schema.env import TOP_LEVEL_KEY
+from spack.schema.env import TOP_LEVEL_KEY, schema, update
```
why change these?  `update` in particular is a very general name that could be shadowed by something here -- I think it's better to leave these namespaces as before.

tgamblin commented on lib/spack/spack/environment/environment.py:
```diff
@@ -2929,93 +2914,25 @@ def included_config_scopes(self) -> List[spack.config.ConfigScope]:
-            msg = "Detected {0} missing include path(s):".format(len(missing))
-            msg += "\n   {0}".format("\n   ".join(missing))
-            raise spack.config.ConfigFileError(msg)
+        # Load config scopes added via environment 'include:' in reverse so
+        # that highest-precedence scopes are last.
+        name_prefix = f"env:{self.name}"
+        scopes: List[spack.config.ConfigScope] = []
+        for entry in reversed(includes):
+            included_scope = spack.config.include_path_scope(
+                spack.config.included_path(entry),
+                name_prefix,
+                self.config_stage_dir,
+                self.manifest_dir,
+            )
+            if included_scope is not None:
+                scopes.append(included_scope)
```
It shouldn't be necessary to do this here if `include:` is handled by the config system.  Just pushing the env's scope should handle any includes.  Similarly, if we ever overrode a scope with `config.override`, it shouldn't have to think about the scope's includes -- any push should just handle includes automatically.

tgamblin commented on lib/spack/spack/error.py:
```diff
@@ -202,3 +202,7 @@ class MirrorError(SpackError):
         super().__init__(msg, long_msg)
+
+
+class RemoteFileError(SpackError):
+    """Problem locating remote file(s)."""
```
If this is only used in one place (remote file cache) prefer to put it there.

tgamblin commented on lib/spack/spack/schema/env.py:
```diff
@@ -29,11 +29,7 @@
-                "specs": spec_list_schema,
-                "include_concrete": include_concrete,
-            },
+            {"specs": spec_list_schema, "include_concrete": include_concrete},
```
hooray! only a couple things left to get out of here (in future PRs)

tgamblin reviewed (changes requested):
Ok I have a number of requests, most of which I've implemented -- I will push that and keep iterating on this later.

With the removals below and by implementing includes in `push_scope`, I got this working for recursive includes, and environments don't need to think much about it at all (which is what we wanted with making `include:` into config)

tgamblin committed:
consolidate include handling to push/pop/remove in config

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
Singleton: allow incremental initialization

Most factories will simply create the object to be initialized
and return it.

In some cases, e.g. when bootstrapping some global state, the
singleton may need to be initialized incrementally. If the factory
returns a generator instead of a regular object, the singleton will
assign each result yielded by the generator to the singleton instance.
This allows methods called by the factory in later stages to refer back
to the singleton.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
Remove web_config now that CONFIG can be initialized incrementally

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
add todo for copying recursive includes on create

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
get rid of a circular dependency

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

haampie commented:
Spack has `sha256` validation almost everywhere, why not here?

tldahlgren commented:
FYI. Running unit tests with the changes made over the weekend, I'm getting some failures (I haven't investigated yet):

```
FAILED lib/spack/spack/test/cmd/env.py::test_init_from_yaml_relative_includes_outside_env
FAILED lib/spack/spack/test/cmd/env.py::test_env_with_included_config_file_url
FAILED lib/spack/spack/test/cmd/env.py::test_env_include_packages_url - Value...
...
FAILED lib/spack/spack/test/env.py::test_env_with_include_defs - ValueError: ...
FAILED lib/spack/spack/test/env.py::test_env_with_include_def_missing - Value...
...
FAILED lib/spack/spack/test/main.py::test_include_cfg - AttributeError: modul...
FAILED lib/spack/spack/test/main.py::test_include_duplicate_source - Attribut...
```

We'll need more unit tests for recursive includes too.

tldahlgren commented:
> Ok I have a number of requests, most of which I've implemented -- I will push that and keep iterating on this later.

So do I assume you're going to finish up the unit tests, etc. for this PR?

tldahlgren committed:
Documentation and docstring refinements

Co-authored-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin commented:
Given that I am back to being swamped with other things I think it would be better if you add the unit tests you mentioned and finish up the PR.  I would not worry about the import issues, and @haampie is going to submit a PR to allow us to circumvent the import checker (as I think this fundamentally adds new cycles that are hard to avoid).

I think the things left to do here are:
* make sure unit tests are passing
* make sure remote includes work
* add a `sha256` field, make it validate, and require it for remote includes:
    ```yaml
    spack:
        include:
            path: https://example.com/foo.yaml
            sha256: "<long hash>"
    ```
* add more unit tests for recursive case
* Think about how diamond includes work -- but save that for last.  @kwryankrattiger mentioned that he wants to ensure that this:
    ```
        a
      /   \
     b     c
      \   /
        d
    ```
    results in both `b` and `c` being able to override `d`.  I think we just need to add a topo-sort to `self.scopes` but that can happen last.  A test case for this would be good, though.

tldahlgren committed:
config.py's include_path_scope: need to pass method, not result

tldahlgren committed:
canonicalize_path: need to recognize and properly handle URL schemes

tldahlgren committed:
Add links to GitHub and GitLab raw file content URLs

tldahlgren committed:
Revised IncludePath comment per feedback

tldahlgren committed:
config: fix comment typo

tldahlgren commented on lib/spack/spack/config.py:
```diff
@@ -774,13 +821,84 @@ def override(
+    if (not include.when) or spack.spec.eval_conditional(include.when):
+        # canonicalize_path does variable expansion and resolves relative paths
+        include_path = spack.util.path.canonicalize_path(include.path)
+        config_path = rfc_util.local_path(include_path, _include_cache_location)
```
Use of `_include_cache_location` is okay for global includes but is not what I thought we'd agreed to for environment includes.

tldahlgren committed:
Rebase cleanup: restore fixture marking to write_config_file

tldahlgren committed:
remote_file_cache: move spack.util.web into functions (NOTE: doesn't break circular import)

tldahlgren committed:
Add mutable_mock_env_path to tests that should not be using user's managed environment

haampie commented:
Trying out this PR:

1. Can you use a hash(url) as a cache filename?
2. If you add sha256 checksum validation can you use that as a cache filename too? If there's a cache hit from sha256 there's no need to warn "will not (re-)fetch file from".
3. These warnings are annoying; it basically warns about a possible cache invalidation issues the second after you've fetched them. For most users that's what they want. If you care about cache invalidation when no sha256 is given, then you could use a generous time-to-live checked against mtime, and fetch after that period without warning at any point in time. Cache invalidation was certainly simpler if sha256 was required, but I guess that's not for everyone.
4. It seems the code path for re-fetching is hit unnecessarily. Why twice here:
   ```
   $ spack -e . config get config
   ==> Warning: Will not (re-)fetch file from ... since it already exists at ....
   ==> Warning: Will not (re-)fetch file from ... since it already exists at ....
   ```

haampie commented on lib/spack/spack/config.py:
```diff
@@ -119,13 +119,27 @@
+INCLUDE_SCOPE_PREFIX = "include"
+
+#: safeguard for recursive includes -- raise an error after this many layers.
+MAX_RECURSIVE_INCLUDES = 100
```
reduce?

tldahlgren replied:
@haampie Not sure what you're asking to be changed here wrt limiting the depth of recursion (or includes). The comment? The variable?

haampie commented on lib/spack/spack/config.py:
```diff
@@ -774,13 +821,85 @@ def override(
+        ConfigFileError: unable to access remote configuration file(s)
+    """
+    # circular dependencies
+    import spack.spec
+    import spack.util.path
+    import spack.util.remote_file_cache as rfc_util
```
Apart from `spack.spec` these can now be moved to top level imports right? You've added inline imports for `spack.util.web` in `spack.util.remote_file_cache`.

As for `spack.spec`, do we really need to have `arch` / `architecture` as a Spec instance for the when condition? Their values are available as strings. That would drop the dependency altogether. Could extend with a `arch_family` string.

tldahlgren replied:
I'll revisit these. Thanks!

tldahlgren commented:
> Trying out this PR:
> 
> 1. Can you use a hash(url) as a cache filename/dirname? Right now it ends up with an arbitrarily deep dir structure. Would also help making `spack config blame` readable, right now it's messy. Also will probably prevent reserved file name related bugs on windows, etc.

That structure was discussed off-line and pre dates your sha256 suggestion.

> 2. If you add sha256 checksum validation can you use that as a cache filename/dirname in that case? If there's a cache hit from sha256 there's no need to warn "will not (re-)fetch file from".

Excellent point, thanks.

> 3. These "no re-fetch warnings" are annoying; it basically warns about a possible cache invalidation the second after you've fetched them. If you care about cache invalidation when no sha256 is given, then you could use a generous time-to-live checked against mtime, and fetch after that period without warning at any point in time. Cache invalidation was certainly simpler if sha256 was required, it surprises me a bit that's not the case.

See my response to 1 above.

> 4. It seems the code path for re-fetching is hit unnecessarily. Why twice here:
>    ```
>    $ spack -e . config get config
>    ==> Warning: Will not (re-)fetch file from ... since it already exists at ....
>    ==> Warning: Will not (re-)fetch file from ... since it already exists at ....
>    ```

The warning exposes pre-existing behavior that has to do with how environments / manifest files are processed.

> 5. I guess it's pre-existing and I've just never seen this, but I don't understand why `collect_urls` is implemented as it is. Why does it involve "spidering". Wouldn't it be much more sensible to attempt downloading `https://example.com/path/{concretizer,config,compilers,packages,repos}.yaml` directly iff the URL ends with `/`? Just like webservers serve e.g. `index.html` by default. (I think it's very much unlike spack to arbitrarily download things from a webserver and just trust it...).

Given the plan is to require the `sha256`, do we even want to allow the URL to be a remote directory?

tldahlgren committed:
Tweak recursive includes depth comment

tldahlgren committed:
Move imports per haampie's feedback

tldahlgren committed:
Use textwrap.wrap to format test_updated_completion_scripts message per tgamblin's feedback

tldahlgren committed:
Fix style issue not previously pushed

tldahlgren committed:
Fix typo

tldahlgren committed:
Fix _ensure_env_dir and changed exception caught by test_init_from_yaml_relative_includes_outside_env

tldahlgren committed:
test/main: Replace removed update_config_with_includes with add_command_line_scopes

tldahlgren committed:
test/bindist.py: move mutable_mock_env_path from all to two tests (will this help Windows?)

tldahlgren committed:
Preliminary sha256 requirement for remote includes (Needs tests)

tldahlgren committed:
Added/revised unit tests for remote_file_cache and remote includes

tldahlgren committed:
Added mock_include_cache so don't pollute user cache's includes

tldahlgren committed:
Fix style issues

tldahlgren committed:
canonicalize_path: ensure normpath run on local files

tgamblin commented:
@greenc-FNAL I think this is getting merged this week or next.

tldahlgren commented:
> @tldahlgren Sorry to be a pain, but what's the current status of this PR, please?

No worries. 🤞 Todd is correct. There is support for recursive includes at this point but the first value encountered when processing includes currently takes precedence.

tldahlgren committed:
canonicalize_path: tweak that hopefully resolves windows test_substitute_config_variables failure

greenc-FNAL commented:
@tldahlgren @tgamblin Thank you!

tldahlgren committed:
test_rfc_local_path_file: fix for windows

tldahlgren committed:
Add test_include_recurse_limit

tldahlgren committed:
WIP: draft include recurse unit test

tldahlgren committed:
Defer ordering of recursive includes to a follow-on PR

tldahlgren committed:
Does this help with mypy error in solver?

tldahlgren committed:
Hopefully fix test_rfc_remote_local_path sha256 check for windows

tldahlgren commented:
@tgamblin Does this need anything else at this point?

marcmengel commented:
This looks excellent to me.  It meets my need of per-os configs, since I can include $os, etc in the paths.
and I could have multiple instances referring to a shared config via URL.

greenc-FNAL commented:
LGTM!

tgamblin commented on lib/spack/spack/test/cmd/env.py:
```diff
@@ -1067,13 +1067,17 @@ def test_init_from_yaml_relative_includes(tmp_path):
         assert os.path.exists(os.path.join(e2.path, f))
 
 
+# TODO: Should we be supporting relative path rewrites when creating new env from existing?
+# TODO: If so, then this should confirm that the absolute include paths in the new env exist.
```
no - we don't support this for views either -- we leave relative paths to outside the env untouched.

tgamblin commented on lib/spack/spack/config.py:
```diff
@@ -763,13 +811,84 @@ def _add_platform_scope(
+    Returns: configuration scope
+
+    Raises:
+        ValueError: included path has an unsupported URL scheme, is required
+            but does not exist; configuration stage directory argument is missing
```
I don't think this is really the right exception type. Before, a non-existing path would get you a `ConfigFileError`, which makes more sense b/c it's something wrong with the config.

[ValueError](https://docs.python.org/3/library/exceptions.html#ValueError) is really for when a *function* gets an invalid type. So I think an invalid URL scheme makes some sense.  But a missing argument doesn't sound like the right thing for a `ValueError`.  I would honestly make all of these `ConfigFileError` so that the caller can tell whether it's a problem with the config file.

tgamblin reviewed (approved):
Looks good! 

I have a few issues left with error types, but I think we should leave that for a later config error overhaul. Specifically:

* I don't think `ValueError` should be coming out of `spack.config`.
* I think *any* error coming out of config files should be raising an error that points the user to the line in the file where things went wrong, but that's a bigger change.

I think we can save the larger fix for the refactor and just get this in as-is.
